\documentclass[]{article}

\usepackage[normalem]{ulem}
\usepackage{xcolor} 
\newcommand{\revch}[1]{{\color{blue} #1}}
\newcommand{\revcom}[1]{{\color{orange} #1}}
\newcommand{\revrem}[1]{{\color{red} \sout{#1}}}
\usepackage{geometry}
\usepackage{hyperref}
\newgeometry{
	top=1.5in,
	bottom=1.5in,
	outer=1.5in,
	inner=1.5in,
}

\usepackage{url}
\usepackage{cuted}
\usepackage{amsmath}
\usepackage{subfig}
%\usepackage{caption}
\usepackage{tabularx}

\usepackage{lipsum}
\usepackage{enumerate}
\usepackage{enumitem}
%load any additional packages
\usepackage{amssymb}
%\usepackage{stackengine}
%\stackMath
\usepackage{amsfonts}
\usepackage{mathtools}

\usepackage{bibentry}
%\usepackage{epstopdf} 
%\usepackage{xr}
%\usepackage[landscape,a4paper]{geometry}
\usepackage{booktabs} 
\usepackage{colortbl} 
%\usepackage{xcolor} 
%\usepackage{xfrac}
%\newcommand{ra}[1]{renewcommand{arraystretch}{#1}}
%\usepackage{pdflscape}
%\usepackage{longtable}
%\usepackage{stackrel}
%\usepackage{hyperref}
%\usepackage[demo]{graphicx}
\usepackage{caption}
\usepackage{subcaption}
%\usepackage{svg}
\usepackage{setspace}
%\usepackage{refcheck}
\usepackage{natbib}
%\usepackage{listings}
%\usepackage{color} 
\renewcommand{\baselinestretch}{1.25}	
%opening
\title{Amino Acid Trait Association Model}
%\author{Michael Golden}
\date{}
\begin{document}

\maketitle

\subsection*{Model description}
We developed a pair of nested evolutionary models, a null and an alternative model, to test for associations between a target amino acid at an alignment site and a target binary trait. The target amino acid being tested, denoted $targetaa$, represents one of the 20 possible amino acids. The target trait, denoted $\mathcal{T}$, represents the trait being tested for an association with the target amino acid. The non-target trait is denoted $\mathcal{N}$. The traits, like the amino acids, are assumed to have evolved in an evolutionary manner.
\iffalse
 and are thus modelled using a two-state continuous-time Markov model akin to Felsenstein's 1981 DNA substitution model \citep{felsenstein1981evolutionary}.
\fi

The null model treats the amino acid evolution at a site and the trait evolution as independent of one another, whereas the alternative model treats the target amino acid and the target binary traits as potentially associated. This potential association is introduced into the alternative model via a dependence parameter $\lambda$. 

For a given amino acid site and set of traits we used maximum likelihood estimation to estimate the parameters of both models and to obtain the maximum likelihood values. The maximum likelihood values were used to compare both models using a likelihood ratio test (LRT) and to calculate a p-value. If the LRT rejects the null model ($p<0.05$) in favour of the alternative model, this suggests that the target amino acid and target trait are associated. 

This association can be a positive association: the target amino acid and the target trait tend to co-occur together, or a negative association: the target amino acid and the target trait tend to actively avoid co-occurring together. When the maximum likelihood estimate for the dependence parameter is larger than one, $\hat{\lambda}>1$, this suggests a positive association, and when $\hat{\lambda}<1$, this suggests a negative association.

This test is somewhat analogous to a chi-squared test of association, except it accounts for phylogenetic correlations. A chi-squared test will treat each observation of amino acid and trait at the tips of a phylogeny as independent events, when in reality they are produced by an evolutionary process where the underlying number of events leading to those observations may be small. This is commonly referred to as a founder effect \citep{bhattacharya2007founder}, and can result in chi-squared associations whose significance is erroneously inflated. Our test avoids this by explicitly modelling the potential dependence between the amino acid and trait evolutionary processes and testing its significance relative to a model that treats them as independent of one another.

Note that the traits, like the amino acids, are assumed to evolve in an evolutionary manner along the tree, and therefore our test is only appropriate where the trait can be described by an evolutionary process. This is the case for the HP and LP traits because they are a direct function of the presence or absence of an insertion, which is generated by an insertion-deletion evolutionary process along the tree. This test would not be appropriate for a trait such as patient survival, which represents a propensity along the tree rather than propagating in a discrete manner. For traits such as these we recommend the test outlined in \citet{bhattacharya2007founder}.

A formal description of the model is given as follows: the joint evolution of amino acids and traits are modelled using a $40\times{40}$ substitution model $Q$, that combines a $20\times{20}$ amino substitution model, $A$, and $2\times{2}$ by trait model, T. The trait model is a two-state continuous-time Markov model akin to Felsenstein's 1981 DNA substitution model \citep{felsenstein1981evolutionary}. The joint model is given as follows:
\begin{equation}
Q_{ij,mn}=\begin{cases}
\mu A_{ij}\lambda & \text{if }i\neq j\text{ and }j=targetaa\text{ and }m=n=\mathcal{T}\\
\mu A_{ij}\frac{1}{\lambda} & \text{if }i\neq j\text{ and }j\neq targetaa\text{ and }m=n=\mathcal{T}\\
\mu A_{ij} & \text{if }i\neq j\text{ and }m=n=\mathcal{N}\\
\pi_{\mathcal{T}}\tau\lambda & \text{if }i=j=targetaa\text{ and }m=\mathcal{N}\text{ and }n=\mathcal{T}\\
\pi_{\mathcal{T}}\tau\frac{1}{\lambda} & \text{if }i=j\text{ and }j\neq target\text{aa and }m=\mathcal{N}\text{ and }n=\mathcal{T}\\
\pi_{\mathcal{N}}\tau & \text{if }i=j\text{ and }m=\mathcal{T}\text{ and }n=\mathcal{N}\\
0 & \text{otherwise}
\end{cases}
\end{equation}

Where $i$ and $j$ represent the initial and end amino acid states, respectively, and $m$ and $n$ represent the start and end trait states, respectively. This formulation is motivated by the RNA base-pairing model of \citet{muse1995evolutionary}.

$\mu$ is a site-specific amino acid substitution rate, and $\tau$ is the trait substitution rate. $A$ is a rate matrix given by the LG2008 substitution model. $\pi^{A}$ is a vector of 20 amino acid frequencies as specified by the LG2008 model, and $\pi_{\mathcal{T}}$ and  $\pi_{\mathcal{N}}$ represents the frequencies of the non-target ($\mathcal{T}$) and target traits ($\mathcal{N}$) states, respectively. 

The equilibrium frequencies, $\pi$, of the alternative model are given by four separate cases corresponding to the two possible values for traits ($\mathcal{T}$ or $\mathcal{N}$), and whether the amino acid ($aa$) matches the target amino acid ($targetaa$) or not:

\begin{equation}
\pi_{aa=targetaa,trait=\mathcal{T}}=k^{-1}\pi_{aa}^{A}\pi_{\mathcal{T}}\lambda
\end{equation}

\begin{equation}
\pi_{aa\neq targetaa,trait=\mathcal{T}}=k^{-1}\pi_{aa}^{A}\pi_{\mathcal{T}}\frac{1}{\lambda}
\end{equation}

\begin{equation}
\pi_{aa=targetaa,trait=\mathcal{N}}=k^{-1}\pi_{aa}^{A}\pi_{\mathcal{N}}
\end{equation}

\begin{equation}
\pi_{aa\neq targetaa,trait=\mathcal{N}}=k^{-1}\pi_{aa}^{A}\pi_{\mathcal{N}}
\end{equation}

Where $\kappa=(\lambda+\frac{1}{\lambda})\pi_{\mathcal{T}}+2\pi_{\mathcal{N}}$ is a normalising constant.

These equilibrium frequencies provide a intuitive way of understanding the influence of the association parameter $\lambda$. It is possible to get a sense of the expected frequencies of particular amino acid and trait associations for given values of $\lambda$. Furthermore, they can be used to predict, for a single sequence, the posterior probability of a trait given the amino $aa$ at the target site:
\begin{equation}
p(trait=\mathcal{T}|aa=targetaa)=\frac{\pi_{\mathcal{T}}\lambda}{\pi_{\mathcal{N}}+\pi_{\mathcal{T}}\lambda}
\end{equation}

\begin{equation}
p(trait=\mathcal{T}|aa\neq targetaa)=\frac{\pi_{\mathcal{T}}}{\pi_{\mathcal{N}}\lambda+\pi_{\mathcal{T}}}
\end{equation}

Also note that the model is time-reversible, and therefore an unrooted tree can be used if the equilibrium probabilities are taken to be the initial probabilities at any rooting of the tree \citep{felsenstein1981evolutionary}.

\subsection*{Simulations}
\begin{table}[h!]
	\centering
	\caption{\label{tab:results} Summary of benchmarks results}	
	\small
	\begin{tabular}{cccc}
		\midrule
		\textbf{Simulated} & \textbf{Simulated} & & \\
		\textbf{association}  & \textbf{rate of trait} & & \\
		\textbf{strength} & \textbf{evolution} & \textbf{Recall} & \textbf{Precision}\\
		\midrule
		Weak (2.0) & 2.0 & 0.23 & 0.94\\
		Intermediate (4.0) & 2.0 & 0.35 & 0.96\\
		Strong (8.0) & 2.0 & 0.50 & 0.93\\
		Weak (2.0) & 4.0 & 0.26 & 0.92\\
		Intermediate (4.0) & 4.0 & 0.48 & 0.92\\
		Strong (8.0) & 4.0 & 0.64 & 0.96\\
		Weak (2.0) & 7.5 & 0.28 & 0.97\\
		Intermediate (4.0) & 7.5 & 0.62 & 0.98\\
		Strong (8.0) & 7.5 & 0.73 & 0.95\\
		\midrule
	\end{tabular}
\end{table}

\begin{figure}[h!]
	\centering
	\includegraphics[width=0.7\columnwidth]{benchmarks_contour.pdf}
	\caption{Contour plot of recall (blue-green-yellow colour gradient) as a function of simulated rate of trait evolution (x-axis) and simulated association strength (y-axis).}
	\label{fig:benchmarks}%
\end{figure}
We simulated amino acid alignments using the empirical H7 AIV maximum likelihood tree and its corresponding HP and LP trait values. Each alignment consisted of 500 amino acids, with the first twenty sites of each alignment simulated as being associated with the traits, each having a different one of the 20 canonical amino acids as the target amino acid. The remaining 480 amino acid sites were simulated under the LG2008 model and were therefore treated as being independent of the traits. Three different degrees of association were simulated: weak, intermediate, and high, combined with three different rates of trait evolution (2.0, 4.0, and 7.0) - the inferred rate of trait evolution in the H7 ML tree was $\sim$4.0 and so this was selected as an intermediate value. 

To account for the potential error introduced during tree inference, an ML tree was inferred using FastTree \citep{price2010fasttree}  for each of the simulated alignments. Potential associations were then estimated using our model on the first 40 sites of each alignment. The first twenty sites were used to measure the number of true positive and false negative detections, whereas the remaining twenty sites were simulated as independent of the traits and were used to measure the number of true negative and false positive detections. The recall and precision were calculated for each simulation using the number of true positives (TN), false-positives (FP), and false-negatives (FN).  Recall and precision are defined as follows:
\begin{equation}
\text{Recall}=\frac{\text{TP}}{\text{TP}+\text{FN}}
\end{equation}

\begin{equation}
\text{Precision}=\frac{\text{TP}}{\text{TP}+\text{FP}}
\end{equation}
The results in Table~\ref{tab:results} indicate that our model has a false-discovery rate (FDR$=100\% \times [1.0 - \text{Precision}]$) of $\sim5\%$ across all test conditions which is consistent with our p-value significance threshold of 0.05. The recall of the model increases with the simulated association strength as expected, with stronger associations being more easily detected. Likewise, the recall of the model increases with a higher rate of trait evolution (Table~\ref{tab:results} and Figure~\ref{fig:benchmarks}), which is also expected given that higher rates of trait evolution imply a greater number of trait events along the tree and therefore more the power in being able to detect an association.

\subsection*{Software availability}
Julia source code (compatible with Windows and Linux) is available at:\\ \href{https://github.com/michaelgoldendev/trait-evolution}{https://github.com/michaelgoldendev/trait-evolution}

\bibliographystyle{natbib}%%%%natbib.sty
\bibliography{refs}%%%refs.bib

\end{document}
